src/modules/utils/directory_win.rs

```diff
@@ -0,0 +1,113 @@
+extern crate winapi;
+
+use std::ffi::OsStr;
+use std::iter;
+use std::mem;
+use std::os::windows::ffi::OsStrExt;
+use winapi::ctypes::c_void;
+use winapi::shared::minwindef::{BOOL, DWORD};
+use winapi::um::handleapi;
+use winapi::um::processthreadsapi;
+use winapi::um::securitybaseapi;
+use winapi::um::winnt::{
+    SecurityImpersonation, DACL_SECURITY_INFORMATION, FILE_ALL_ACCESS, FILE_GENERIC_EXECUTE,
+    FILE_GENERIC_READ, FILE_GENERIC_WRITE, GENERIC_MAPPING, GROUP_SECURITY_INFORMATION, HANDLE,
+    OWNER_SECURITY_INFORMATION, PRIVILEGE_SET, PSECURITY_DESCRIPTOR, STANDARD_RIGHTS_READ,
+    TOKEN_DUPLICATE, TOKEN_IMPERSONATE, TOKEN_QUERY,
+};
+
+pub fn is_write_allowed(folder_path: &str) -> std::result::Result<bool, &'static str> {
+    let folder_name: Vec<u16> = OsStr::new(folder_path)
+        .encode_wide()
+        .chain(iter::once(0))
+        .collect();
+    let mut length: DWORD = 0;
+
+    let rc = unsafe {
+        securitybaseapi::GetFileSecurityW(
+            folder_name.as_ptr(),
+            OWNER_SECURITY_INFORMATION | GROUP_SECURITY_INFORMATION | DACL_SECURITY_INFORMATION,
+            std::ptr::null_mut(),
+            0,
+            &mut length,
+        )
+    };
+    if rc != 0 {
+        return Err(
+            "GetFileSecurityW returned non-zero when asked for the security descriptor size",
+        );
+    }
+
+    let mut buf: Vec<u8> = Vec::with_capacity(length as usize);
+
+    let rc = unsafe {
+        securitybaseapi::GetFileSecurityW(
+            folder_name.as_ptr(),
+            OWNER_SECURITY_INFORMATION | GROUP_SECURITY_INFORMATION | DACL_SECURITY_INFORMATION,
+            buf.as_mut_ptr() as *mut c_void,
+            length,
+            &mut length,
+        )
+    };
+
+    if rc != 1 {
+        return Err("GetFileSecurityW failed to retrieve the security descriptor");
+    }
+
+    let mut token: HANDLE = 0 as HANDLE;
```

Comment:
Clearly, we need to create an impersonation token here, but I'm having trouble understanding why. Is this because `AccessCheck` is a much more powerful API than just checking for file permissions, and so needs an impersonation token for other uses of it? Is this something that you just have to do when you want to call `AccessCheck`?

Reply:
AccessCheck requires an impersonation token (per https://docs.microsoft.com/en-us/windows/win32/api/securitybaseapi/nf-securitybaseapi-accesscheck).
You can find some sample code on how to do call DuplicateToken here: 
https://github.com/jeanga/wil-rs/blob/abc43b2d3ec45896ebf2f3d19b380e7bfed085e6/src/token.rs#L139

HTH